src/tribol/geom/NodalNormal.cpp

```diff
@@ -0,0 +1,192 @@
+      for ( int d = 0; d < mesh_view.spatialDimension(); ++d ) {
+        mesh_view.getNodalNormals()( d, nodeId ) += mesh_view.getElementNormals()( d, i );
+      }
+
+    }  // end loop over element nodes
+
+  }  // end loop over elements
+
+  // normalize the nodal normals
+  if ( mesh_view.spatialDimension() == 3 ) {
+    for ( int i = 0; i < mesh_view.numberOfNodes(); ++i ) {
+      RealT mag = magnitude( mesh_view.getNodalNormals()( 0, i ), mesh_view.getNodalNormals()( 1, i ),
+                             mesh_view.getNodalNormals()( 2, i ) );
+      mesh_view.getNodalNormals()( 0, i ) /= mag;
+      mesh_view.getNodalNormals()( 1, i ) /= mag;
+      mesh_view.getNodalNormals()( 2, i ) /= mag;
+    }
+  } else {
+    for ( int i = 0; i < mesh_view.numberOfNodes(); ++i ) {
+      RealT mag = magnitude( mesh_view.getNodalNormals()( 0, i ), mesh_view.getNodalNormals()( 1, i ) );
+      mesh_view.getNodalNormals()( 0, i ) /= mag;
+      mesh_view.getNodalNormals()( 1, i ) /= mag;
+    }
+  }
+}
+
+#ifdef TRIBOL_USE_ENZYME
+
+void VertexAvgNormal::Compute( MeshData& mesh )
+{
+  SLIC_ERROR_ROOT_IF( mesh.spatialDimension() != 3, "3D mesh required for vertex averaged normal." );
+
+  mesh.allocateNodalNormals();
+
+  auto n0 = Array2D<RealT>( { 3, mesh.numberOfNodes() }, mesh.getAllocatorId() );
+  n0.fill( 0.0 );
+
+  if ( compute_deriv_ ) {
+    getJacobianData().reserveBlockJ( { BlockSpace::NONMORTAR }, mesh.numberOfElements() );
+  }
+
+  auto mesh_view = mesh.getView();
+
+  SLIC_ERROR_IF( !mesh_view.hasReferencePosition(),
+                 "Reference coordinates must be registered for vertex averaged normal." );
+
+  auto num_nodes_per_elem = mesh_view.numberOfNodesPerElement();
+  for ( int e{ 0 }; e < mesh_view.numberOfElements(); ++e ) {
+    RealT x[12];
+    RealT xref[12];
+    RealT n[12];
+    for ( int i{ 0 }; i < num_nodes_per_elem; ++i ) {
+      int node_id = mesh_view.getGlobalNodeId( e, i );
+      for ( int d{ 0 }; d < 3; ++d ) {
+        x[d * num_nodes_per_elem + i] = mesh_view.getPosition()[d][node_id];
+        xref[d * num_nodes_per_elem + i] = mesh_view.getReferencePosition()[d][node_id];
+        n[d * num_nodes_per_elem + i] = 0.0;
+      }
+    }
+    if ( compute_deriv_ ) {
+      StackArray<DeviceArray2D<RealT>, 9> blockJ( 3 );
+      blockJ( 0, 0 ) = DeviceArray2D<RealT>( 12, 12 );
+      blockJ( 0, 0 ).fill( 0.0 );
+      ElementVertexAvgNormalJacobian( x, xref, n, blockJ( 0, 0 ).data(), num_nodes_per_elem );
+      getJacobianData().storeElemBlockJ( { e }, blockJ );
+    } else {
+      ElementVertexAvgNormal( x, xref, n, num_nodes_per_elem );
+    }
+    // assemble normal contribution
+    for ( int i{ 0 }; i < num_nodes_per_elem; ++i ) {
+      int node_id = mesh_view.getGlobalNodeId( e, i );
+      for ( int d{ 0 }; d < 3; ++d ) {
+        mesh_view.getNodalNormals()( d, node_id ) += n[d * num_nodes_per_elem + i];
+      }
+    }
+    // compute reference normal
```

Comment:
I think once I understand the approach for the calculation this reference normal stuff will make more sense.